File: ui/web-app/src/domain/cloudProviderRuntime.ts
// SPDX-FileCopyrightText: 2026 Aerobag contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import type {
  CloudHttpRequest,
  CloudHttpResponse,
} from "../generated/cloudWire";

export async function executeCloudHttpRequest(
  request: CloudHttpRequest,
): Promise<CloudHttpResponse> {
  try {
    const headers = new Headers(request.headers.map((header) => [header.name, header.value]));
    const response = await fetch(request.url, {
      method: request.method.toUpperCase(),
      headers,
      body: request.body_base64 == null ? undefined : decodeBase64Url(request.body_base64),
    });
    const declaredLength = Number(response.headers.get("content-length"));
    if (Number.isFinite(declaredLength) && declaredLength > request.max_response_bytes) {
      await response.body?.cancel();
      return { result: "response_too_large", limit_bytes: request.max_response_bytes };
    }
    const bytes = await readBoundedBody(response, request.max_response_bytes);
    if (bytes === null) {
      return { result: "response_too_large", limit_bytes: request.max_response_bytes };
    }
    return {
      result: "completed",
      status_code: response.status,
      body_base64: encodeBase64Url(bytes),
    };
  } catch (error) {
    return {
      result: "transport_error",
      detail: error instanceof Error ? error.message : String(error),
    };
  }
}

async function readBoundedBody(response: Response, limit: number): Promise<Uint8Array | null> {
  if (!response.body) {
    return new Uint8Array();
  }
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }
    length += value.byteLength;
    if (length > limit) {
      await reader.cancel();
      return null;
    }
    chunks.push(value);
  }
  const bytes = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    bytes.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return bytes;
}

function decodeBase64Url(value: string): ArrayBuffer {
  const padded = value.replaceAll("-", "+").replaceAll("_", "/").padEnd(Math.ceil(value.length / 4) * 4, "=");
  const binary = atob(padded);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0)).buffer;
}

function encodeBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (let offset = 0; offset < bytes.length; offset += 0x8000) {
    binary += String.fromCharCode(...bytes.subarray(offset, offset + 0x8000));
  }
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "");
}

File: ui/android-app/app/src/main/java/org/aerobag/app/CloudProviderRuntime.kt
// SPDX-FileCopyrightText: 2026 Aerobag contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package org.aerobag.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Headers
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okio.Buffer
import org.aerobag.app.generated.CloudHttpMethod
import org.aerobag.app.generated.CloudHttpRequest
import org.aerobag.app.generated.CloudHttpResponse

internal class AndroidCloudProviderRuntime(
    private val httpClient: OkHttpClient = OkHttpClient(),
) {
    suspend fun executeHttp(request: CloudHttpRequest): CloudHttpResponse =
        executeCloudHttpRequest(request, httpClient)
}

internal suspend fun executeCloudHttpRequest(
    request: CloudHttpRequest,
    httpClient: OkHttpClient,
): CloudHttpResponse = withContext(Dispatchers.IO) {
    try {
        val headers = Headers.Builder().apply {
            request.headers.forEach { add(it.name, it.value) }
        }.build()
        val contentType = request.headers
            .firstOrNull { it.name.equals("content-type", ignoreCase = true) }
            ?.value
            ?.toMediaTypeOrNull()
        val requestBody = when {
            request.bodyBase64 != null -> decodeBase64Url(request.bodyBase64).toRequestBody(contentType)
            request.method == CloudHttpMethod.Post || request.method == CloudHttpMethod.Put ->
                ByteArray(0).toRequestBody(contentType)
            else -> null
        }
        val call = Request.Builder()
            .url(request.url)
            .headers(headers)
            .method(
                when (request.method) {
                    CloudHttpMethod.Get -> "GET"
                    CloudHttpMethod.Post -> "POST"
                    CloudHttpMethod.Put -> "PUT"
                    CloudHttpMethod.Delete -> "DELETE"
                },
                requestBody,
            )
            .build()
        httpClient.newCall(call).execute().use { response ->
            val bytes = readBoundedResponse(response.body, request.maxResponseBytes)
                ?: return@withContext CloudHttpResponse.ResponseTooLarge(request.maxResponseBytes)
            CloudHttpResponse.Completed(
                statusCode = response.code,
                bodyBase64 = encodeBase64Url(bytes),
            )
        }
    } catch (error: Throwable) {
        CloudHttpResponse.TransportError(error.message ?: error::class.simpleName ?: "HTTP request failed")
    }
}

private fun readBoundedResponse(body: okhttp3.ResponseBody?, limit: Long): ByteArray? {
    if (body == null) return ByteArray(0)
    if (body.contentLength() > limit) return null
    val source = body.source()
    val buffer = Buffer()
    while (true) {
        val read = source.read(buffer, 8_192L)
        if (read == -1L) break
        if (buffer.size > limit) return null
    }
    return buffer.readByteArray()
}

private fun decodeBase64Url(value: String): ByteArray =
    java.util.Base64.getUrlDecoder().decode(value)

private fun encodeBase64Url(bytes: ByteArray): String =
    java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
